### PR TITLE
[objc] Add Array support

### PR DESCRIPTION
**Pull request by dalexsoto**
ReturnTypes Support
- [x] byte [] to NSData
- [x] string [] to NSArray<NSString *> *
- [x] Foo [] to NSArray<Foo *> *
- [x] numeric valueType [] to NSArray<NSNumber *> *
- [x] Parameters Support

### CODE REVIEW

File: objcgen/objcgenerator.cs [hunk 1]
```diff
@@ -796,8 +796,124 @@ protected override void Generate (ProcessedMethod method)
...
+				implementation.WriteLine ("unsigned int __resarrval = mono_array_get (__resarr, unsigned int, __residx);");
+				implementation.WriteLine ("__resobj = [NSNumber numberWithUnsignedInt:__resarrval];");
+				break;
+			case TypeCode.UInt64:
+				implementation.WriteLine ("unsigned long long __resarrval = mono_array_get (__resarr, unsigned long long, __residx);");
+				implementation.WriteLine ("__resobj = [NSNumber numberWithUnsignedLongLong:__resarrval];");
+				break;
+			case TypeCode.Byte:
+				implementation.WriteLine ("unsigned char __resarrval = mono_array_get (__resarr, unsigned char, __residx);");
+				implementation.WriteLine ("[__resobj appendBytes:&__resarrval length:1];");
```
Comment (spouliot) [inline objcgen/objcgenerator.cs]: That will need to be optimized (byte arrays can be very large). You should be able to skip the loop using `mono_array_addr` and the array length, e.g.

```
[__resobj appendBytes:mono_array_addr (__resarr) length:__resarrlength];
```

File: objcgen/objcgenerator.cs [hunk 2]
```diff
@@ -796,8 +796,124 @@ protected override void Generate (ProcessedMethod method)
...
+				break;
+			default:
+				throw new NotImplementedException ($"Converting type {t.Name} to a native type name");
+			}
+
+			if (typecode == TypeCode.Byte) {
+				implementation.WriteLine ("}");
+				implementation.WriteLine ("return __resobj;");
+			}
+			else {
```
Comment (spouliot) [inline objcgen/objcgenerator.cs]: minor/style: `} else {`

File: objcgen/objcgenerator.cs [hunk 3]
```diff
@@ -796,8 +796,124 @@ protected override void Generate (ProcessedMethod method)
...
+		void ReturnArrayValue (Type t)
+		{
+			var typecode = Type.GetTypeCode (t);
+			implementation.WriteLine ("MonoArray* __resarr = (MonoArray *) __result;");
+			implementation.WriteLine ("int __resarrlength = mono_array_length (__resarr);");
+
+			if (typecode == TypeCode.Byte)
+				implementation.WriteLine ("NSMutableData* __resobj = [NSMutableData dataWithCapacity:__resarrlength];");
+			else {
+				implementation.WriteLine ("id __autoreleasing* __resarrbuf = (id __autoreleasing *) malloc (sizeof (id) * __resarrlength);");
```
Comment (spouliot) [inline objcgen/objcgenerator.cs]: any reason for `__autoreleasing` ?
ARC wants something for I would have used `__strong` since you're doing the allocation (it's not a variable that some other code will allocate) and it's usage is all local

File: objcgen/objcgenerator.cs [hunk 4]
```diff
@@ -796,8 +796,124 @@ protected override void Generate (ProcessedMethod method)
...
 		}
 
+		void ReturnArrayValue (Type t)
+		{
+			var typecode = Type.GetTypeCode (t);
+			implementation.WriteLine ("MonoArray* __resarr = (MonoArray *) __result;");
+			implementation.WriteLine ("int __resarrlength = mono_array_length (__resarr);");
+
+			if (typecode == TypeCode.Byte)
+				implementation.WriteLine ("NSMutableData* __resobj = [NSMutableData dataWithCapacity:__resarrlength];");
```
Comment (spouliot) [inline objcgen/objcgenerator.cs]: It would be best to return a non-mutable `NSData` as is match what the managed code is doing, e.g.

```
NSData *__resobj = [NSData dataWithBytes:mono_array_addr (__resarr) length: __resarrlength];
```

That would also avoid the loop (see my other comment).

File: objcgen/NameGenerator.cs [hunk 5]
```diff
@@ -146,6 +149,32 @@ public static string GetMonoName (Type t)
...
+			case TypeCode.UInt64:
+				return "NSArray <NSNumber *>";
+			case TypeCode.Byte:
+				return "NSData";
+			case TypeCode.String:
+				return "NSArray <NSString *>";
+			case TypeCode.Object:
+				return $"NSArray<{GetObjCName (t)} *>";
+			default:
+				throw new NotImplementedException ($"Converting type {t.Name} to a native type name");
```
Comment (chamons) [inline objcgen/NameGenerator.cs]: Should this throw a:

```throw ErrorHelper.CreateError (99, "Internal error```

Comment (spouliot) [inline objcgen/NameGenerator.cs]: we catch any `NotImplementedException` and throw an error

Comment (chamons) [inline objcgen/NameGenerator.cs]: Then why do we throw 99's specifically in places? Maybe we should either just throw NotImplementedException or throw 99's?

Comment (spouliot) [inline objcgen/NameGenerator.cs]: A `99` is something that should not happen (or be reached) and for which we need a test case.

A `NotImplementedException` message is for incomplete code, the exception message should be enough (without a test case).

Comment (chamons) [inline objcgen/NameGenerator.cs]: Ahhhh. Now I'm enlightened.

👍

File: objcgen/objcgenerator.cs [hunk 6]
```diff
@@ -796,8 +796,124 @@ protected override void Generate (ProcessedMethod method)
 				GenerateDefaultValuesWrappers (objcsig, method.Method);
 		}
 
+		void ReturnArrayValue (Type t)
+		{
+			var typecode = Type.GetTypeCode (t);
+			implementation.WriteLine ("MonoArray* __resarr = (MonoArray *) __result;");
+			implementation.WriteLine ("int __resarrlength = mono_array_length (__resarr);");
```
Comment (rolfbjarne) [inline objcgen/objcgenerator.cs]: What if `__resarr` is NULL?

Comment (dalexsoto) [inline objcgen/objcgenerator.cs]: Should we just return nil? Good catch

Comment (spouliot) [inline objcgen/objcgenerator.cs]: 👍 let's add a test that returns `null` from managed :)

**Comment by dalexsoto** [comment]
Reversing to don't merge, I might need some bits from this hopefully can have a working version soonish, but reviews to the current code are welcome, return type should be done

File: objcgen/objcgenerator.cs [hunk 7]
```diff
@@ -799,8 +799,127 @@ protected override void Generate (ProcessedMethod method)
...
+				implementation.WriteLine ("__resobj = [NSNull null];");
+				implementation.Indent--;
+				break;
+			case TypeCode.Object:
+				var tname = NameGenerator.GetTypeName (t);
+				if (types.HasProtocol (t))
+					tname = "__" + tname + "Wrapper";
+				implementation.WriteLine ("MonoObject* __resarrval = mono_array_get (__resarr, MonoObject *, __residx);");
+				implementation.WriteLine ($"{tname}* __tmpobj = [[{tname} alloc] initForSuper];");
+				implementation.WriteLine ("if (__resarrval) {");
```
Comment (rolfbjarne) [inline objcgen/objcgenerator.cs]: You're creating `__tmpobj` even if `__resarrval` is NULL, you can move the `__tmpobj` creation inside the `if` to avoid this.

File: tests/managed/arrays.cs [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+		public short [] ShortArrMethod () => new short [] { short.MaxValue, short.MinValue, 0 };
+		public int [] IntArrMethod () => new [] { int.MaxValue, int.MinValue, 0 };
+		public long [] LongArrMethod () => new [] { long.MaxValue, long.MinValue, 0 };
+		public ushort [] UshortArrMethod () => new ushort [] { ushort.MaxValue, ushort.MinValue, 10 };
+		public uint [] UintArrMethod () => new uint [] { uint.MaxValue, uint.MinValue, 15 };
+		public ulong [] UlongArrMethod () => new ulong [] { ulong.MaxValue, ulong.MinValue, 117 };
+		public byte [] ByteArrMethod () => new byte[] { 0x0, 0x01, 0x02, 0x04, 0x08 };
+
+		public string [] GetNull { get; } = null;
+		public string [] StringArr { get; } = new [] { "Hola", "Hello", "Bonjour" };
```
Comment (rolfbjarne) [inline tests/managed/arrays.cs]: Also try adding a `null` entry in the string array.

**Comment by dalexsoto** [comment]
@spouliot @rolfbjarne Done with feedback also added parameters support 😄

File: support/mono-support.h [hunk 9]
```diff
@@ -47,6 +47,20 @@ MonoString *    mono_object_to_string (MonoObject *obj, MonoObject **exc);
 MonoClass *     mono_class_get (MonoImage *image, uint32_t type_token);
 MonoMethod *    mono_get_method (MonoImage *image, uint32_t token, MonoClass *klass);
 MonoClassField* mono_class_get_field (MonoClass *klass, uint32_t field_token);
+MonoClass*      mono_get_string_class(void);
```
Comment (spouliot) [inline support/mono-support.h]: minor: mono style puts a space before the `(`

File: tests/objc-cli/libmanaged/Tests/Tests.m [hunk 10]
```diff
@@ -783,6 +783,307 @@ - (void)testNestedClasses {
 	XCTAssert ([nested additionX:2 y:4] == 6, "Nested class method call");
 }
 
+- (void)testArraySupport {
+	Arrays_Arr *testClass = [[Arrays_Arr alloc] init];
+
+	NSArray<NSString *> *strArr = [testClass stringArrMethod];
+	XCTAssertEqual([strArr count], 3, @"strArr count");
```
Comment (spouliot) [inline tests/objc-cli/libmanaged/Tests/Tests.m]: minor: mono style puts a space before the `(`

File: objcgen/objcgenerator-helpers.cs [hunk 11]
```diff
@@ -54,5 +54,43 @@ void GetSignatures (string objName, string monoName, MemberInfo info, ParameterI
...
 			monoSignature = mono.ToString ();
 		}
+
+		public static string GetArrayCreator (string parameterName, Type type)
+		{
+			string arrayCreator = $"MonoArray* __{parameterName}arr = mono_array_new(__mono_context.domain, {{0}}, __{parameterName}length);";
+
+			switch (Type.GetTypeCode (type)) {
+			case TypeCode.String:
+				return string.Format (arrayCreator, "mono_get_string_class()");
```
Comment (spouliot) [inline objcgen/objcgenerator-helpers.cs]: minor: mono style puts a space before the `(`

File: objcgen/objcgenerator-helpers.cs [hunk 12]
```diff
@@ -54,5 +54,43 @@ void GetSignatures (string objName, string monoName, MemberInfo info, ParameterI
 			objcSignature = objc.ToString ();
 			monoSignature = mono.ToString ();
 		}
+
+		public static string GetArrayCreator (string parameterName, Type type)
+		{
+			string arrayCreator = $"MonoArray* __{parameterName}arr = mono_array_new(__mono_context.domain, {{0}}, __{parameterName}length);";
```
Comment (spouliot) [inline objcgen/objcgenerator-helpers.cs]: minor: mono style puts a space before the `(`

**Comment by dalexsoto** [comment]
@spouliot thanks! Done 👍
